modules/standard/Lists.chpl

```diff
@@ -0,0 +1,868 @@
+
+      :throws IllegalArgumentError: If the given element cannot be found.
+    */
+    proc indexOf(x: eltType, start: int=1, end: int=size): int throws {
+      _enter();
+
+      try _boundsCheckLeaveOnThrow(start);
+      try _boundsCheckLeaveOnThrow(end);
+
+      for i in start..end do
+        if x == _getRef(i) then {
+          _leave();
+          return i;
+        }
+
+      _leave();
+
+      const msg = "No such element in List: " + x:string;
+      throw new owned
+        IllegalArgumentError(msg);
+
+      // Should never reach here.
+      return -1;
+    }
+
+    /*
+      Return the number of times a given element is found in this List.
+
+      :arg x: An element to count.
+
+      :return: The number of times a given element is found in this List.
+      :rtype: `int`
+    */
+    proc count(x: eltType): int {
+      _enter();
+
+      var result = 0;
+
+      for i in 1.._size do {
+        if x == _getRef(i) then
+          result += 1;
+      }
+
+      _leave();
+      return result;
+    }
+
+    /*
+       Sort the elements of this List in place using their default sort order.
+
+       .. warning::
+
+        Sorting the contents of this List may invalidate existing references
+        to the elements contained in this List.
+    */
+    proc sort() {
+      // TODO: Need to look in the Sort module to see the API.
+    }
+
+    /*
+      Sort the items of this List in place using a comparator.
+
+      .. warning::
+
+        Sorting the elements of this List may invalidate existing references
+        to the elements contained in this List.
+
+      :arg comparator: A comparator object used to sort this List.
+    */
+    proc sort(comparator) {
+      // TODO: Need to look in the Sort module to see the API. 
+    }
+
+    /*
+      Index this List via subscript. Returns a reference to the element at a
+      given index in this List.
+
+      :arg i: The index of the element to access.
+
+      :return: An element from this List.
+
+      :throws IllegalArgumentError: If the given index is out of bounds.
+    */
+    proc const this(i: int) ref throws {
+      _enter();
+
+      //
+      // TODO: Do we want to switch to array style indexing instead?
+      //
+      if false && !_boundsCheck(i) then {
+        _leave();
+        const msg = "Bad index into List: " + i:string;
+        halt(msg);
+      }
+
+      try _boundsCheckLeaveOnThrow(i);
+      ref result = _getRef(i);
+      _leave();
+      return result;
+    }
+
+    /*
+      Produce a serial iterator over the elements of this List.
+
+      :yields: A reference to one of the elements contained in this List.
+    */
+    iter these() ref {
+
+      //
+      // TODO: I'm not even sure of what the best way to WRITE a threadsafe
+      // iterator is, _let alone_ whether it should even be threadsafe
+      // (I mean, is there even a point when reference/iterator invalidation
+      // is still a thing?
```

Comment:
Hmm.. I'm not sure either, but I think you should be able to merge the PR with this draft.

Additionally I think you'll want leader/follower and standalone parallel iterators.